### DISCUSSION TITLE
Support smart contract program

### DISCUSSION THREAD

**Discussion by masstensor** [veracruz-project/veracruz#245] [Ideas]
Can Veracruz support WASM based smart contract?

**Comment by dominic-mulligan-arm** [comment]
Hi @masstensor, and thanks for your interest in Veracruz.  First, I'll admit I'm not familiar enough with how a Wasm-based smart contract works, to be sure.  Having said that:

Veracruz can support a wide range of Wasm programs in support of privacy-preserving distributed computations.  Specifically, any distributed computation that can be expressed as a collection of data providers, pairing with a program provider, to compute a private computation over the input data sets can be handled.

So, if the smart contract itself can be isolated as a standalone Wasm binary, and all of its required inputs be fed into Veracruz, then yes, we can probably support that.

**Reply by masstensor** [reply]
Hi Dominic,

Thank for your reply.
I am sorry for my unclearly idea description, in fact, I want to know that whether Veracruz can support Ethereum's gas liked mechanism to prevent a confidential computing program infinitely-executing. With this mechanism running the confidential smart contract program in Veracruz platform will be possible. 

Thanks.

**Reply by dominic-mulligan-arm** [reply]
Hi @masstensor,

OK, thanks for clarifying!  At present we have no support for that, but this is something we have discussed in the past, and from memory we were in the process of adding support for this in an earlier stage of the project.

One difficulty here is that Veracruz supports multiple execution strategies: both interpretation and JITing of Wasm.  Were we to add support for this we would first need to define a common cost model that remained well-defined with respect to the execution strategy.  We could do this by assigning a "cost" to Wasm byte code instructions themselves, I suppose, but the book-keeping associated with tracking this would also need to be reflected in the generated binary code spat out by our JIT (Wasmtime).

Moreover, what should that cost be?  Just a uniform cost model, with every instruction consuming 1 unit of gas, or something more sophisticated e.g. multiplications taking 2 units, whilst loads only taking 1?  I think if your goal is to prevent non-termination of the program, then a uniform cost model suffices.

**Reply by masstensor** [reply]
OK, I understood, thank for your reply
